### DISCUSSION TITLE
Compiling the library unsuccesfull

### DISCUSSION THREAD

**Discussion by mcsmit** [redboltz/async_mqtt#358] [Q&A] [answered]
When I try to build the library on Windows with the options ASYNC_MQTT_BUILD_EXAMPLES and ASYNC_MQTT_USE_TLS it fails with many errors. I load the library in vscode and build with CMake selecting VS2022 amd64 as kit. Is building possible in Windows?

Then I tried to build on Linux. With the same options, building does start but needs a lot of memory. Building fails after a long time (10 minutes?) due to a memory shortage. Just before quitting, it's using all 14 GB (6 GB RAM and 8 GB swap). Is this normal behaviour? I use cmake and gcc 12 in Linux.

Please let me know if more information is needed.

**Comment by redboltz** [comment, marked as answer]
For msvc, I guess that you need to find OpenSSL using cmake.
https://cmake.org/cmake/help/v3.25/module/FindOpenSSL.html

Linux case, for example compiling `cl_cpp17_mqtt_pub`, memory consumption is about 1.2GB.

```
/usr/bin/time -v make cl_cpp17_mqtt_pub   
[  0%] Building CXX object example/CMakeFiles/cl_cpp17_mqtt_pub.dir/cl_cpp17_mqtt_pub.cpp.o
[100%] Linking CXX executable cl_cpp17_mqtt_pub
[100%] Built target cl_cpp17_mqtt_pub
        Command being timed: "make cl_cpp17_mqtt_pub"
        User time (seconds): 24.27
        System time (seconds): 1.37
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:25.78
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1238156    <=== here. about 1.2GB
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 325311
        Voluntary context switches: 55
        Involuntary context switches: 105
        Swaps: 0
        File system inputs: 0
        File system outputs: 218760
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

You can check memory consumption like this.

In example, there are two subdirectories.

```
example/separate_client
example/separate_endpoint
```
They are examples for https://redboltz.github.io/async_mqtt/doc/latest/separate.html

It requires a log of memories. If you don't need this mode, you can just skip it.

**Reply by mcsmit** [reply]
OpenSSL is found correctly on Windows but the build failed on various other compile errors. Then I tried to build using clang-cl and that gave a better result. Then only the C++20 examples build failed so I disabled them and the build succeeded. I guess using cl to build is the problem.
I have to try your Linux suggestions yet.

**Reply by mcsmit** [reply]
For Linux, I first had to find out how to use your suggestion. I found out that in vscode I needed to select the kit 'unspecified' so it created Makefile files. My build used even less memory: 1.9 GB. Thanks for your help so far.

**Reply by redboltz** [reply]
It seems that you Linux issue has been fixed. It is good. I wll add cmake option to enable example/separate_* build and set OFF by default.

For windows, the following CI (for MSVC) would help you.
https://github.com/redboltz/async_mqtt/blob/main/.github/workflows/gha.yml#L175-L177
ASYNC_MQTT_USE_TLS is not enabled, but it could still useful e.g.) $env:CL="/D_WIN32_WINNT#0x0A00 /DBOOST_THREAD_VERSION#3 /DBOOST_ASIO_NO_DEPRECATED /bigobj /EHsc /Zc:preprocessor"

Windows build setting is complecated because the default setting is not support pure C++20.

**Reply by mcsmit** [reply]
Thanks for the extra information, I will give that a try. In the meantime, I found my memory problem. I ran my builds in vscode using the cmake extension. And this combination seems to result in memory problems. I will try to run outside vscode.
